vhost-user-backend/Cargo.toml

```diff
@@ -9,11 +9,13 @@ edition = "2021"
 license = "Apache-2.0"
 
 [features]
+default = ["userfaultfd"]
 xen = ["vm-memory/xen", "vhost/xen"]
 
 [dependencies]
 libc = "0.2.39"
 log = "0.4.17"
+userfaultfd = { version = "0.7.0", optional = true }
```

Comment:
Just another question (more for the user of this crate, like `virtiofsd` and `vhost-device`). Since we are adding a new dependency, and maybe some backends don't need the `postcopy` feature, should we gate this with a new feature (enabled by default)?

Just to avoid some unneeded dependency during packaging.

@germag @vireshk @epilys @stsquad @Ablu WDYT?

Reply:
As of now `userfaultfd` crate is optional and enabled by default. So if users don't want the unneeded dependency they can specify the crate with `vhost = { ..., default-features=false }`. 

Or I can switch to gate postcopy with separate feature which is not enabled by default (so it will be like `xen` feature). Then I can update 
```rust
#[cfg(not(feature = "xen"))]
```
 into 
```rust
#[cfg(all(feautre = "postcopy", not(feature = "xen")))]
```
Hm, I think second option should be the best here.

Reply:
> As of now `userfaultfd` crate is optional and enabled by default. So if users don't want the unneeded dependency they can specify the crate with `vhost = { ..., default-features=false }`.
> 

Right, but what happens if in the future we will add another feature, and we want to enable only `postcopy`?

> Or I can switch to gate postcopy with separate feature which is not enabled by default (so it will be like `xen` feature). Then I can update
> 
> ```rust
> #[cfg(not(feature = "xen"))]
> ```
> 
> into
> 
> ```rust
> #[cfg(all(feautre = "postcopy", not(feature = "xen")))]
> ```
> 
> Hm, I think second option should be the best here.

Yeah, I think so. Or have some kind of build failure if both are enabled.

Reply:
> Yeah, I think so. Or have some kind of build failure if both are enabled.

Here there is an example: https://doc.rust-lang.org/std/macro.compile_error.html

Reply:
Ok, I updated the code, but now the CI is in the way. It runs tests and clippy with `--all-features` so it fail now.

Reply:
> Ok, I updated the code, but now the CI is in the way. It runs tests and clippy with `--all-features` so it fail now.

That should be fixed, because testing with `--all-features` i'st effectively _only_ testing the 'xen' feature, since that feature disables the 'default' feature in vm-memory(*)

(*)  vm-memory should be also fixed, because it allows to compile it with '--all-fetures' and then failing in runtime

Reply:
Do I understand correctly that you are saying that `rust-vmm-ci` and `vm-memory` should be fixed? (The 'That' part of the first sentence is confusing) Otherwise I don's see how this crate can do a work around for this issue.

Reply:
> Do I understand correctly that you are saying that `rust-vmm-ci` and `vm-memory` should be fixed? (The 'That' part of the first sentence is confusing) Otherwise I don's see how this crate can do a work around for this issue.

yep, in `rust-vmm-ci`. One workaround we can do is to define our custom `.buildkite` pipeline.
But this is for sure for another PR. I can do that, but tomorrow I'll be off the whole day, so I can't do before Friday.

Reply:
> Just another question (more for the user of this crate, like `virtiofsd` and `vhost-device`). Since we are adding a new dependency, and maybe some backends don't need the `postcopy` feature, should we gate this with a new feature (enabled by default)?

I got no opinion here. I assume userfaultfd is just a shallow wrapper around fd operations? So I assume it does not blow up the code size by much. So I may also be fine with keeping it if adding knobs becomes a maintenance burden too.

Reply:
@Ablu yes, I see your point, but it definitely carries other dependencies, and when we go packaging (for example in fedora), it's always better to have as few dependencies as possible, especially if we know we're not going to use that feature.

Also, in this case we need gates for that code anyway, so it seems free to me.

Reply:
I posted this: https://github.com/rust-vmm/vhost/pull/229

@ShadowCurse can you rebase on it, and add `postcopy` feature in `.buildkite/rust-vmm-ci-tests.json` in `*-all-without-xen` tests?

`rust-vmm-ci` pipeline is still active, so it should fail, but `custom-rust-vmm-ci` should work.
I'll deactivate it when we merge #229

Reply:
Done

Reply:
thanks! It looks like it is working good! I'll rebuild the pipeline when I'll disable the old pipeline

Reply:
Looking at this I think there are 3 approaches we can use to fix this:

1. In vhost wrap `lib.rs` in `cfg(not(all(feature="xen", feature="postcopy")))` so when compiling with `xen` and `postcopy` it compiles an empty crate where `cargo build` and `cargo test` will work.
2. Update rust-vmm-ci to not use `--all-features` so it only tests the default features by default and crates can add additional tests for their feature sets.
3. Do https://github.com/rust-vmm/rust-vmm-ci/issues/152

I think solutions 1 and 2 are both viable and reasonably easy to implement. Both 2 and 3 will affect other crates. I think 3 raises a few too many concerns about how much it increase runtime of tests. I would probably recommend going with 1 for now as its the easiest to implement since it doesn't require external changes.

Reply:
> Looking at this I think there are 3 approaches we can use to fix this:

@JonathanWoollett-Light thanks for your suggestions!
I proposed #229 as another approach, maybe we should continue the discussion there.

> 
>     1. In vhost wrap `lib.rs` in `cfg(not(all(feature="xen", feature="postcopy")))` so when compiling with `xen` and `postcopy` it compiles an empty crate where `cargo build` and `cargo test` will work.

Does this mean that our CI will always go to test an empty crate?

> 
>     2. Update rust-vmm-ci to not use `--all-features` so it only tests the default features by default and crates can add additional tests for their feature sets.

We have PRs (like this one for example) where almost all the code is behind a feature and so CI would not run `clippy` for example on this new code. I don't know if that's good.

> 
>     3. Do [Add `cargo-all-features` rust-vmm-ci#152](https://github.com/rust-vmm/rust-vmm-ci/issues/152)

IMHO we should move toward this direction, though, somewhat limiting the testing matrix and providing only a set of features to test. So using it more to avoid testing unsupported combinations.

Reply:
> Looking at this I think there are 3 approaches we can use to fix this:
> 
>     1. In vhost wrap `lib.rs` in `cfg(not(all(feature="xen", feature="postcopy")))` so when compiling with `xen` and `postcopy` it compiles an empty crate where `cargo build` and `cargo test` will work.
> 
>     2. Update rust-vmm-ci to not use `--all-features` so it only tests the default features by default and crates can add additional tests for their feature sets.
> 
>     3. Do [Add `cargo-all-features` rust-vmm-ci#152](https://github.com/rust-vmm/rust-vmm-ci/issues/152)
> 
> 
> I think solutions 1 and 2 are both viable and reasonably easy to implement. Both 2 and 3 will affect other crates. I think 3 raises a few too many concerns about how much it increase runtime of tests. I would probably recommend going with 1 for now as its the easiest to implement since it doesn't require external changes.

I disagree with 1, silently creating an empty crate it's not better than current situation. Enabling xen with an incompatible feature must be a compilation error, instead of a runtime one. This is also a problem for some distributions that recommend to build crates with `--all-features` (like Arch)

Reply:
> > ```
> > 2. Update rust-vmm-ci to not use `--all-features` so it only tests the default features by default and crates can add additional tests for their feature sets.
> > ```
> 
> We have PRs (like this one for example) where almost all the code is behind a feature and so CI would not run `clippy` for example on this new code. I don't know if that's good.

Okay, I had misinterpreted, yes I think this is a viable option, then we have to do something like #229 (with just some tests included there).